File: androidApp/src/main/java/com/mbta/tid/mbta_app/android/util/fetchApi.kt
package com.mbta.tid.mbta_app.android.util

import android.util.Log
import com.mbta.tid.mbta_app.model.response.ApiResult
import com.mbta.tid.mbta_app.repositories.IErrorBannerStateRepository

/**
 * If `getData` returns an `ApiResultOk`, clears any preexisting data error in `errorKey` in
 * `errorBannerRepo` and calls `onSuccess`. If `getData` returns an `ApiResultError` or throws, sets
 * an error in `errorKey` in `errorBannerRepo` with the given `onRefreshAfterError`.
 */
suspend fun <T : Any> fetchApi(
    errorBannerRepo: IErrorBannerStateRepository,
    errorKey: String,
    getData: suspend () -> ApiResult<T>,
    onSuccess: suspend (T) -> Unit = {},
    onRefreshAfterError: () -> Unit,
) {
    val result: ApiResult<T> =
        try {
            getData()
        } catch (e: Exception) {
            ApiResult.Error(code = null, message = e.message ?: "")
        }
    when (result) {
        is ApiResult.Error -> {
            Log.e("fetchApi", "API request $errorKey failed: $result")
            errorBannerRepo.setDataError(
                key = errorKey,
                details = result.toString(),
                action = onRefreshAfterError,
            )
        }
        is ApiResult.Ok -> {
            errorBannerRepo.clearDataError(key = errorKey)
            onSuccess(result.data)
        }
    }
}

File: iosApp/iosApp/Utils/FetchApi.swift
//
//  FetchApi.swift
//  iosApp
//
//  Created by Horn, Melody on 2024-09-30.
//  Copyright © 2024 MBTA. All rights reserved.
//

import Shared

/// If `getData` returns an `ApiResultOk`, clears any preexisting data error in `errorKey` in `errorBannerRepo` and
/// calls `onSuccess`.
/// If `getData` returns an `ApiResultError` or throws, sets an error in `errorKey` in `errorBannerRepo` with the given
/// `onRefreshAfterError`.
/// If `getData` is interrupted by task cancellation, does nothing.
func fetchApi<T>(
    _ errorBannerRepo: IErrorBannerStateRepository = RepositoryDI().errorBanner,
    errorKey: String,
    getData: () async throws -> ApiResult<T>,
    onSuccess: @MainActor (T) -> Void = { _ in },
    onRefreshAfterError: @escaping () -> Void
) async {
    var result: ApiResult<T>
    do {
        result = try await getData()
    } catch is CancellationError {
        return
    } catch {
        result = ApiResultError<T>(code: nil, message: error.localizedDescription)
    }
    switch onEnum(of: result) {
    case let .ok(result):
        errorBannerRepo.clearDataError(key: errorKey)
        await onSuccess(result.data)
    case let .error(error):
        errorBannerRepo.setDataError(key: errorKey, details: error.description(), action: onRefreshAfterError)
    }
}
